baselines/flwr_baselines/publications/adaptive_federated_optimization/cifar/client.py

```diff
@@ -0,0 +1,157 @@
+import ray
+import torch
+from flwr.common.typing import Scalar, Weights
+from torch.utils.data import DataLoader
+
+from .utils import ClientDataset, get_cifar_model, get_transforms, test, train
+
+
+class RayClient(fl.client.NumPyClient):
+    """Ray Virtual Client."""
+
+    def __init__(self, cid: str, fed_dir: Path, num_classes: int):
+        """Implements Ray Virtual Client.
+
+        Args:
+            cid (str): Client ID, in our case a str representation of an int.
+            fed_dir (Path): Path where partitions are saved.
+            num_classes (int): Number of classes in the classification problem.
+        """
+        self.cid = cid
+        self.fed_dir = fed_dir
+        self.num_classes = num_classes
+        self.properties: Dict[str, Scalar] = {"tensor_type": "numpy.ndarray"}
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    def get_parameters(self) -> Weights:
+        """Returns weight from a given model. If no model is passed, then a
+        local model is created. This can be used to initialize a model in the
+        server.
+
+        Returns:
+            Weights: weights from the model.
+        """
+        net = get_cifar_model(self.num_classes)
+        weights = [val.cpu().numpy() for _, val in net.state_dict().items()]
+        # parameters = weights_to_parameters(weights)
+        return weights
+
+    def get_properties(self, config: Dict[str, Scalar]) -> Dict[str, Scalar]:
+        """Returns properties for this client.
+
+        Args:
+            config (Dict[str, Scalar]): Options to be used for selecting specific
+            properties.
+
+        Returns:
+            Dict[str, Scalar]: Returned properties.
+        """
+        return self.properties
+
+    def fit(
+        self, parameters: Weights, config: Dict[str, Scalar]
+    ) -> Tuple[Weights, int, Dict[str, Scalar]]:
+        """Usual fit function that performs training locally.
+
+        Args:
+            parameters (Weights): Initial set of weights sent by the server.
+            config (Dict[str, Scalar]): coonfig file containing num_epochs,etc...
+
+        Returns:
+            Tuple[Weights, int, Dict[str, Scalar]]: New set of weights,
+            number of samples and dictionary of metrics.
+        """
+        net = self.set_parameters(parameters)
+        net.to(self.device)
+        num_workers = len(ray.worker.get_resource_ids()["CPU"])
+        trainset = ClientDataset(
+            path_to_data=Path(self.fed_dir) / f"{self.cid}" / "train.pt",
+            transform=get_transforms(self.num_classes)["train"],
+        )
+
+        trainloader = DataLoader(
+            trainset, batch_size=int(config["batch_size"]), num_workers=num_workers
+        )
+        # train
+        train(net, trainloader, epochs=int(config["epochs"]), device=self.device)
+
+        # return local model and statistics
+        weights = [val.cpu().numpy() for _, val in net.state_dict().items()]
```

Comment:
why not use `self. get_parameters()`?

Reply:
Hum... Where exactly? `get_parameters` I think is mainly used to initialize weights on the server-side.

Reply:
ok, like you have it looks good